pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -826,7 +826,30 @@ func getCRICadvisorStats(infos map[string]cadvisorapiv2.ContainerInfo) map[strin
 		if !isPodManagedContainer(&info) {
 			continue
 		}
-		stats[path.Base(key)] = info
+		stats[extractIDFromCgroupPath(key)] = info
 	}
 	return stats
 }
+
+func extractIDFromCgroupPath(cgroupPath string) string {
+	//case0: /kubepods/burstable/pod2fc932ce-fdcc-454b-97bd-aadfdeb4c340/9be25294016e2dc0340dd605ce1f57b492039b267a6a618a7ad2a7a58a740f32
+	id := path.Base(cgroupPath)
+
+	//case1: "/kubepods-burstable-pod2fc932ce_fdcc_454b_97bd_aadfdeb4c340.slice:cri-containerd:db622e6f7b16079db56fd7814031f7e8fc78b3050c24fe2fc431beeebbeb1c8e"
```

Comment:
I'm still not personally convinced we need this case

Reply:
- I am sure for the case1: `"/kubepods-burstable-pod2fc932ce_fdcc_454b_97bd_aadfdeb4c340.slice:cri-containerd:db622e6f7b16079db56fd7814031f7e8fc78b3050c24fe2fc431beeebbeb1c8e"` . 
    - Just did a live bug, here's the piece of output
```
   157:         allInfos, err := getCadvisorContainerInfo(p.cadvisor)
-> 158:         if err != nil {
   159:                 return nil, fmt.Errorf("failed to fetch cadvisor stats: %v", err)
   160:         }
   161:         caInfos := getCRICadvisorStats(allInfos)
   162:
(dlv) p allInfos
map[string]k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerInfo [
        "/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod50e56a8b_4a44_48d4_a1d8_f2a088697230.slice": *{
                Spec: (*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerSpec")(0xc0019959e0),
                Stats: []*k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats len: 2, cap: 2, [
                        *(*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats")(0xc001237860),
                        *(*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats")(0xc001237950),
                ],}, 
        "/kubepods-besteffort-pod172f50f4_10ce_4f06_af33_3e1f0716b2a0.slice:cri-containerd:fa98e6c3491302833bd4b68c9ca866569cc40e7858d5881501e84b4dad29e3bb": *{
                Spec: (*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerSpec")(0xc00148a7e0),
                Stats: []*k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats len: 2, cap: 2, [
                        *(*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats")(0xc0023d4000),
                        *(*"k8s.io/kubernetes/vendor/github.com/google/cadvisor/info/v2.ContainerStats")(0xc0023d40f0),
                ],}, 
```

- But I am not sure for the case2 since I cannot reproduce it anymore. (But I did notice it when debugging)
```
/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod2fc932ce_fdcc_454b_97bd_aadfdeb4c340.slice/cri-containerd-aaefb9d8feed2d453b543f6d928cede7a4dbefa6a0ae7c9b990dd234c56e93b9.scope
```

Reply:
well, you convinced me :smile:

Reply:
I tried to repro and print out `allInfos` in my environment and I'm getting examples of case2 (i.e. `/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podcd388b79_4555_4c44_8a68_f67b00538b93.slice/cri-containerd-6c0ab3f65ff9f6c4641efc26ea2ba5c6ffac7d1de684a6251bcd901c5eb6aa12.scope`) etc

Here is full dump: https://gist.githubusercontent.com/bobbypage/a701361188b53c8f34a7bcfdbdf71908/raw/ceba53a77820a948d34b908f1ff1ad2d300ea304/debug.txt

This is with both kubelet configured to use systemd cgroup driver as well as containerd configured to use systemd cgroup driver. 

I'm curious in what setup you got case1 (i.e. how is kubelet / containerd configured in terms of cgroup driver?). Is there a mismatch of driver usuage, i.e. kubelet using cgroupfs driver but containerd using systemd cgroup driver or vice versa in that case?

Reply:
For case1, both kubelet and containerd are configured to use systemd cgroup in my test. 
My environment is kubernetes 1.21.2 + containerd 1.4.6. 
And also I double checked
1. the kubelet log, it shows systemd cgroup driver.
```
container_manager_linux.go:283] "Creating Container Manager object based on Node Config" nodeConfig={RuntimeCgroupsName: SystemCgroupsName: KubeletCgroupsName: ContainerRuntime:remote CgroupsPerQOS:true CgroupRoot:/ CgroupDriver:systemd 
```
2. for the containerd log, there's no error log about cgroup. Here's my configuration
```
version = 2

[plugins]
  [plugins."io.containerd.grpc.v1.cri"]
    sandbox_image = xx
    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
    runtime_type = "io.containerd.runc.v2"
  [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
    SystemdCgroup = true
```
but the `crictl info | grep system` shows
```
$ crictl info | grep system
    "systemdCgroup": false,
```
I am not sure whether containerd is really using the systemd_cgroup driver.

Reply:
It might be due to that in my environment, the kubelet and containerd process is not managed by systemd. 
And I tried in another environment with systemd managing kubelet and containerd, I got same cgroup path as you @bobbypage 

So I just removed the case1 check.`kubepods-besteffort-pod172f50f4_10ce_4f06_af33_3e1f0716b2a0.slice:cri-containerd:0db8ad3f044cd5b67a1be535368e804a5df42ae`